### DISCUSSION TITLE
Retrieve signed data to verify the digital signature using endesive

### DISCUSSION THREAD

**Discussion by MallRoy** [pymupdf/PyMuPDF#2135] [Looking for help] [answered]
Is there a way to retrieve signed data using pymupdf and verify the digital signature using endesive. Sample verification code can be found at https://github.com/m32/endesive/blob/master/endesive/pdf/verify.py

**Comment by JorjMcKie** [comment, marked as answer]
This feature is currently not supported in PyMuPDF.
But you can locate the respective data - very much like endesive - using PyMuPDF's low-level API.
A draft for the approach can be found in #2078.

**Reply by MallRoy** [reply]
is there a way to detect the xref of the signature object?

**Reply by JorjMcKie** [reply]
First identify the signature annotation, take its xref. Then extract the value of the `/V` key in that object.
This is either a dictionary or another xref pointing to a dictionary.
One of the keys of the /V dictionary is `/ByteRange`. The other one the respective `/Content`.

Example: a signed PDF. We are looking up the signature field and extract  the above values:
```python 
doc = fitz.open("signed.pdf")
for page in doc:
    for widget in page.widgets(types=[fitz.PDF_WIDGET_TYPE_SIGNATURE]):
        xref = widget.xref
        value = doc.xref_get_key(xref, "V")
        if value[0] == "xref":
            vxref = int(value[1].split()[0])  # xref of the V-field
            print(doc.xref_get_keys(vxref))  # the keys of this PDF dictionary

('ByteRange', 'Contents', 'Filter', 'M', 'Name', 'Prop_Build', 'SubFilter', 'Type')

            brange = (doc.xref_get_key(vxref, "ByteRange"))
            print(brange)

('array', '[0 671901 681503 829736]')

```
This should cover the usual case.
Let me know if - unexpectedly - `/V` is not an xref. Then you must do some additional stuff.
Presumably you know how to interpret the array `[0 671901 681503 829736]`?
